### PR TITLE
refactor(linter): add `declare_all_lint_rules` proc macro

### LINKED ISSUE #113: refactor(linter): use proc_macro to for rule registration
[This file](https://github.com/Boshen/oxc/blob/main/crates/oxc_linter/src/rules.rs) is getting huge and it is just repetitive boilerplate. It would be nice to use a proc_macro to generate these.

The file should end up with something like 

```
declare_all_lints! {
    no_debugger,
    typescript::no_empty_interface,
    deepscan::uninvoked_array_callback
}
```

The identifier `deepscan::uninvoked_array_callback` should be a `syn::Path` I think.

It is expanded into

```
mod deepscan {
  pub mod uninvoked_array_callback;
}
```

For example 

https://github.com/Boshen/oxc/blob/4829b85c2d89cfc8241e8542e49f4f9a8f8d36ae/crates/oxc_linter/src/rules.rs#L5-L7

has [this directory structure](https://github.com/Boshen/oxc/tree/4829b85c2d89cfc8241e8542e49f4f9a8f8d36ae/crates/oxc_linter/src/rules)

### PR DESCRIPTION
**Pull request by amitdahan**
# Summary
Closes #113

### CODE REVIEW

File: crates/oxc_macros/src/declare_all_lint_rules.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        Ok(Self { paths })
+    }
+}
+
+fn define_rule_mod(path: &syn::Path) -> TokenStream {
+    let mut segments = path.segments.iter().rev().peekable();
+    let first = &segments.next().unwrap().ident;
+    let mut stmts = quote! {mod #first;};
+    if segments.peek().is_some() {
```
Comment (YoniFeng) [inline crates/oxc_macros/src/declare_all_lint_rules.rs:25]: very nit:
is there a need for the duplication?

**Review by YoniFei — APPROVED** [review]
LGTM

File: crates/oxc_macros/src/declare_all_lint_rules.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_macros/src/declare_all_lint_rules.rs:25]: we can reduce more code if it's doable

**Comment by Boshen** [comment]
The macro implementation will end up with something like:

```rust
quote! {
    #(#mod_name)*
    #(#use_name)*

    lazy_static::lazy_static! {
        pub static ref RULES: Vec<RuleEnum> = vec![
            ...
        ];
    }


    pub enum RuleEnum {
        #(#struct_name(#struct_name)),*
    }

    impl ...

}
```

It'll be easier if you parse things into a struct, and them implement utils on the struct, e.g.

```rust
pub struct AllLints {
    lints: ...
}

impl AllLints {
    pub fn mod_names ...
}

pub struct SingleLint {
    module: Option<Ident>,
    ident: Ident,
}

impl SingleLint {
    fn mod_name(&self) -> Ident {
        self.ident.clone()
    }

    ...
}

```

```rust
impl Parse for AllLints {
    fn parse(input: ParseStream) -> Result<Self> {
        let list = Punctuated::<SingleLint, Token![,]>::parse_terminated(input)?;
        ...
    }
```

File: crates/oxc_macros/src/declare_all_lint_rules.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+pub fn declare_all_lint_rules(metadata: AllLintRulesMeta) -> TokenStream {
+    let AllLintRulesMeta { rules } = metadata;
+
+    let mod_stmts = rules.iter().map(|rule| rule.mod_stmt());
+    let use_stmts = rules.iter().map(|rule| rule.use_stmt());
+    let names = rules.iter().map(|rule| &rule.name);
+    let names2 = names.clone();
+    let names3 = names.clone();
+    let names4 = names.clone();
+    let names5 = names.clone();
```
Comment (amitdahan) [inline crates/oxc_macros/src/declare_all_lint_rules.rs:83]: This is pretty terrible, but I honestly wasn't sure how to do better 😓

File: crates/oxc_macros/src/declare_all_lint_rules.rs (hunk 1 above)
Comment (amitdahan) [reply crates/oxc_macros/src/declare_all_lint_rules.rs]: > very nit:
> is there a need for the duplication?

Not nitpicking at all, this bothered me, I looked for a cleaner way to achieve:
- `pub mod` for nested `mod` statements
- `;` at the end of the inner most `mod` statement
- no `pub` for the top-level `mod` statement

Somehow this is the best I could come up with, very open to suggestions!

**Comment by amitdahan** [comment]
Pushed the extraction of the rest of the boilerplate.

Since it actually does what it intends to do this time, I'd love to know if anyone thinks the code can be improved any since I'm not so happy with how it turned out.

File: crates/oxc_macros/src/declare_all_lint_rules.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_macros/src/declare_all_lint_rules.rs:83]: Should this be called `struct_names`? And I don't think you need to clone it, `syn` should be able to use it repetitively (I think).

Comment (Boshen) [reply crates/oxc_macros/src/declare_all_lint_rules.rs:83]: Oh! you just need to collect `names` into a Vec!

File: crates/oxc_macros/src/declare_all_lint_rules.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+impl LintRuleMeta {
+    pub fn mod_stmt(&self) -> TokenStream {
+        let mut segments = self.path.segments.iter().rev().peekable();
+        let first = &segments.next().unwrap().ident;
+        let mut stmts = quote! {mod #first;};
+        if segments.peek().is_some() {
+            stmts = quote! {pub #stmts};
+        }
+
+        while let Some(segment) = segments.next() {
```
Comment (Boshen) [inline crates/oxc_macros/src/declare_all_lint_rules.rs:21]: I'm not sure if we are ever going to get 3 levels deep 🤔

**Comment by Boshen** [comment]
> since I'm not so happy with how it turned out.

Which part? This is exact what we need :-)

Quick background: we are using static dispatch here for better readibiliy and performance. We could use a `Box<dyn Rule>`, but I don't really like having `dyn` in a deterministic system.

**Comment by amitdahan** [comment]
> Which part? This is exact what we need :-)

Oh mostly the way the code ended up looking, not the actual goal that it indeed achieved 😅

File: crates/oxc_macros/src/declare_all_lint_rules.rs (hunk 2 above)
Comment (amitdahan) [reply crates/oxc_macros/src/declare_all_lint_rules.rs]: oh wow I'm dumb

**Review by Boshen — APPROVED** [review]

**Review by Boshen — CHANGES_REQUESTED** [review]
Last Conflict :-(

**Comment by Boshen** [comment]
Awesome work! This is going to help a lot once we pump rules out really quickly.

**Comment by shannonrothe** [comment]
@Boshen should we update the `rule_generator` template to take advantage of the new macro?

**Pull request merged**